Factor out some test helpers in hydro

carmaza opened the pull request:
## Proposed changes

Several tests in hydro systems require random fields with physical meaning, e.g. the Lorentz factor must be greater or equal than one, and must be consistent with a spatial velocity. Rather than having a bunch of different calls to make_with_random_values with fine-tuned distributions for each variables, this PR factors out some duplicated functions that can be reused in other tests across hydro systems and pointwise functions.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe commented on tests/Unit/PointwiseFunctions/Hydro/TestHelpers.hpp:
```diff
@@ -0,0 +1,64 @@
+
+// Needed in some tests, so we put it here for convenience
+template <size_t Dim, typename DataType>
+tnsr::ii<DataType, Dim> random_spatial_metric(
```
This one (and maybe `random_lorentz_factor` and `random_velocity`) is more general than just hydro.  We should move it to somewhere in GR, but that can be put off until someone wants it if you want.

carmaza replied:
I agree. Maybe the tests in `tests/Unit/PointwiseFunctions/GeneralRelativity/TestHelpers.hpp` can be reworked similarly in the future.

wthrowe reviewed (commented):
You have some clang-tidy errors, but the gcc test failure looks spurious.  (I'll verify the failure locally and file a bug once my computer finishes what it is doing.)

carmaza committed:
Add test helpers for physically acceptable random tensors

carmaza committed:
Add test helper for unit form along coordinate axes

carmaza committed:
Use test helpers in Valencia

carmaza committed:
Use test helpers in ValenciaDivClean

carmaza committed:
Use test helpers in Hydro

carmaza commented:
I rewrote the history a bit so it is easier to see which helpers I'm adding (I added a few more in the process) and where I am using them. I'm fine with moving any functions to a better location (either for this PR or in the future), just let me know where.

wthrowe reviewed (dismissed):
The 3+1 stuff should be moved into `tests/Unit/PointwiseFunctions/GeneralRelativity/TestHelpers.hpp`, but there's a lot of other stuff in that file that also should be cleaned up, so let's do that in another PR.

nilsdeppe commented on tests/Unit/PointwiseFunctions/Hydro/TestHelpers.cpp:
```diff
@@ -0,0 +1,192 @@
+Scalar<DataType> random_density(const gsl::not_null<std::mt19937*> generator,
+                                const DataType& used_for_size) noexcept {
+  // 1 g/cm^3 = 1.62e-18 in geometric units
+  constexpr double minimum_density = 1.0e-5;
```
Why did you choose such a large value for the minimum?

nilsdeppe commented on tests/Unit/PointwiseFunctions/Hydro/TestHelpers.cpp:
```diff
@@ -0,0 +1,192 @@
+  // assumes Ideal gas with gamma = 4/3
+  // For ideal fluid T = (m/k_b)(gamma - 1) epsilon
+  // where m = atomic mass unit, k_b = Boltzmann constant
+  return Scalar<DataType>{3.21e-3 *
```
Could you make it clear in the comment where this number comes from?

nilsdeppe commented on tests/Unit/PointwiseFunctions/Hydro/TestHelpers.cpp:
```diff
@@ -0,0 +1,192 @@
+// See LICENSE.txt for details.
+
+#include "tests/Unit/PointwiseFunctions/Hydro/TestHelpers.hpp"
+
```
Need `<random>` for `uniform_real_distribution`

`<cstddef>`

nilsdeppe commented on tests/Unit/PointwiseFunctions/Hydro/Test_TestHelpers.cpp:
```diff
@@ -0,0 +1,44 @@
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/Gsl.hpp"
+#include "tests/Unit/PointwiseFunctions/Hydro/TestHelpers.hpp"
```
you need the include for `square` too, no?

nilsdeppe reviewed (changes requested):
I'm happy for you to squash right away

carmaza commented:
Squashed and rebased on develop.

kidder commented:
@wthrowe please merge this if you are happy with it
